File: src/baseline.test.ts
// AST baseline tests, in the style of the TypeScript compiler. Each .java file
// under __fixtures__/cases is parsed and bound, the resulting tree + diagnostics
// are serialized to a stable text form, and compared to the checked-in baseline
// under __fixtures__/baselines.
//
// To (re)generate baselines after an intentional change:
//   UPDATE_BASELINES=1 node --run test
// then review the diff before committing.

import { test } from "node:test";
import { expect } from "expect";
import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { bindSourceFile } from "./binder.ts";
import { forEachChild, parseSourceFile } from "./parser.ts";
import { syntaxKindToString } from "./utilities.ts";
import {
  type Diagnostic,
  type Identifier,
  type LiteralExpression,
  type Node,
  NodeFlags,
  SyntaxKind,
} from "./types.ts";

const here = dirname(fileURLToPath(import.meta.url));
const casesDir = join(here, "__fixtures__", "cases");
const baselinesDir = join(here, "__fixtures__", "baselines");
const shouldUpdate = process.env.UPDATE_BASELINES === "1";

function flagSuffix(flags: NodeFlags): string {
  const parts: string[] = [];
  if (flags & NodeFlags.ThisNodeHasError) parts.push("HasError");
  if (flags & NodeFlags.ThisNodeOrAnySubNodesHasError) parts.push("SubtreeHasError");
  return parts.length ? ` (${parts.join(", ")})` : "";
}

function nodeLabel(node: Node): string {
  let label = syntaxKindToString(node.kind);
  if (node.kind === SyntaxKind.Identifier) {
    label += ` "${(node as Identifier).text}"`;
  } else if (
    node.kind === SyntaxKind.NumericLiteral ||
    node.kind === SyntaxKind.StringLiteral ||
    node.kind === SyntaxKind.CharacterLiteral ||
    node.kind === SyntaxKind.TextBlockLiteral
  ) {
    label += ` ${JSON.stringify((node as LiteralExpression).value)}`;
  }
  return `${label} [${node.pos},${node.end}]${flagSuffix(node.flags)}`;
}

function printNode(node: Node, depth: number, out: string[]): void {
  out.push("  ".repeat(depth) + nodeLabel(node));
  forEachChild(node, child => {
    printNode(child, depth + 1, out);
    return undefined;
  });
}

function formatDiagnostics(title: string, diagnostics: readonly Diagnostic[]): string[] {
  if (diagnostics.length === 0) return [];
  return ["", `${title}:`, ...diagnostics.map(d => `  [${d.pos},${d.end}] ${d.messageText}`)];
}

function serialize(fileName: string, source: string): string {
  const sf = parseSourceFile(fileName, source);
  bindSourceFile(sf);
  const out: string[] = [];
  printNode(sf, 0, out);
  out.push(...formatDiagnostics("Parse diagnostics", sf.parseDiagnostics));
  out.push(...formatDiagnostics("Bind diagnostics", sf.bindDiagnostics ?? []));
  return out.join("\n") + "\n";
}

const cases = existsSync(casesDir)
  ? readdirSync(casesDir)
      .filter(f => f.endsWith(".java"))
      .sort()
  : [];

for (const file of cases) {
  test(`baseline: ${file}`, () => {
    const source = readFileSync(join(casesDir, file), "utf8");
    const actual = serialize(file, source);
    const baselinePath = join(baselinesDir, file.replace(/\.java$/, ".txt"));

    if (shouldUpdate || !existsSync(baselinePath)) {
      mkdirSync(baselinesDir, { recursive: true });
      writeFileSync(baselinePath, actual);
    }
    const expected = readFileSync(baselinePath, "utf8");
    expect(actual).toBe(expected);
  });
}

File: togo/internal/compiler/baseline_test.go
package compiler

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// AST baseline tests, in the style of the TypeScript compiler. Each .java file
// under test-fixtures/parser/cases is parsed and bound, the resulting tree +
// diagnostics serialized to a stable text form, and compared to the checked-in
// baseline under test-fixtures/parser/baselines. Port of src/compiler/
// baseline.test.ts; the Go serialization must match the TS-generated baselines
// byte-for-byte, proving the parser produces an identical tree shape.

func flagSuffix(flags NodeFlags) string {
	var parts []string
	if flags&NodeFlagThisNodeHasError != 0 {
		parts = append(parts, "HasError")
	}
	if flags&NodeFlagThisNodeOrAnySubNodesError != 0 {
		parts = append(parts, "SubtreeHasError")
	}
	if len(parts) == 0 {
		return ""
	}
	return " (" + strings.Join(parts, ", ") + ")"
}

// jsonString reproduces JS JSON.stringify(value) for a string (no HTML escaping).
func jsonString(s string) string {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(s)
	return strings.TrimRight(buf.String(), "\n")
}

func nodeLabel(node *Node) string {
	label := node.Kind.String()
	switch node.Kind {
	case Identifier:
		label += " \"" + node.AsIdentifier().Text + "\""
	case NumericLiteral, StringLiteral, CharacterLiteral, TextBlockLiteral:
		label += " " + jsonString(node.AsLiteralExpression().Value)
	}
	return label + " [" + itoaDiff(node.Pos) + "," + itoaDiff(node.End) + "]" + flagSuffix(node.Flags)
}

func printNode(node *Node, depth int, out *[]string) {
	*out = append(*out, strings.Repeat("  ", depth)+nodeLabel(node))
	node.ForEachChild(func(child *Node) bool {
		printNode(child, depth+1, out)
		return false
	})
}

func formatBaselineDiagnostics(title string, diagnostics []Diagnostic) []string {
	if len(diagnostics) == 0 {
		return nil
	}
	out := []string{"", title + ":"}
	for _, d := range diagnostics {
		out = append(out, "  ["+itoaDiff(d.Pos)+","+itoaDiff(d.End)+"] "+d.MessageText)
	}
	return out
}

func serializeTree(fileName, source string) string {
	sf := ParseSourceFile(fileName, source)
	BindSourceFile(sf)
	sfd := sf.AsSourceFile()
	var out []string
	printNode(sf, 0, &out)
	out = append(out, formatBaselineDiagnostics("Parse diagnostics", sfd.ParseDiagnostics)...)
	out = append(out, formatBaselineDiagnostics("Bind diagnostics", sfd.BindDiagnostics)...)
	return strings.Join(out, "\n") + "\n"
}

func TestParserBaselines(t *testing.T) {
	casesDir := filepath.Join("..", "..", "..", "test-fixtures", "parser", "cases")
	baselinesDir := filepath.Join("..", "..", "..", "test-fixtures", "parser", "baselines")
	entries, err := os.ReadDir(casesDir)
	if err != nil {
		t.Skip("no parser cases present")
	}
	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".java") {
			continue
		}
		t.Run(e.Name(), func(t *testing.T) {
			source, err := os.ReadFile(filepath.Join(casesDir, e.Name()))
			if err != nil {
				t.Fatal(err)
			}
			actual := serializeTree(e.Name(), string(source))
			baselinePath := filepath.Join(baselinesDir, strings.TrimSuffix(e.Name(), ".java")+".txt")
			expected, err := os.ReadFile(baselinePath)
			if err != nil {
				t.Fatalf("no baseline %s: %v", baselinePath, err)
			}
			if actual != string(expected) {
				t.Errorf("AST serialization differs from the baseline.\n--- got ---\n%s\n--- want ---\n%s", actual, expected)
			}
		})
	}
}
